core/templates/services/playthrough.service.ts

```diff
@@ -1,334 +1,334 @@
-  }
-
-  private hasRecordingFinished(): boolean {
-    return (
-      this.hasRecordingBegun() &&
-      this.recordedLearnerActions.length > 1 &&
-      this.recordedLearnerActions[this.recordedLearnerActions.length - 1]
-        .actionType === AppConstants.ACTION_TYPE_EXPLORATION_QUIT);
-  }
-
-  private isRecordedPlaythroughHelpful(): boolean {
-    return (
-      // Playthroughs are only helpful in their entirety.
-      this.hasRecordingFinished() &&
-      // Playthroughs are only helpful if learners have attempted an answer.
-      this.recordedLearnerActions.some(
-        a => a.actionType === AppConstants.ACTION_TYPE_ANSWER_SUBMIT) &&
-      // Playthroughs are only helpful if learners have invested enough time.
-      this.playthroughDurationInSecs >=
-          ServicesConstants.MIN_PLAYTHROUGH_DURATION_IN_SECS);
-  }
-}
-
-angular.module('oppia').factory(
-  'PlaythroughService',
-  downgradeInjectable(PlaythroughService));
+// Copyright 2020 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Service for recording and scrutinizing playthroughs.
+ */
+
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { Injectable } from '@angular/core';
+
+import { AppConstants } from 'app.constants';
+import { ExplorationFeaturesService } from
+  'services/exploration-features.service';
+import {
+  CyclicStateTransitionsCustomizationArgs,
+  EarlyQuitCustomizationArgs,
+  MultipleIncorrectSubmissionsCustomizationArgs
+} from 'domain/statistics/PlaythroughIssueObjectFactory';
+import { LearnerAction, LearnerActionObjectFactory } from
+  'domain/statistics/LearnerActionObjectFactory';
+import { Playthrough, PlaythroughObjectFactory } from
+  'domain/statistics/PlaythroughObjectFactory';
+import { PlaythroughBackendApiService } from
+  'domain/statistics/playthrough-backend-api.service';
+import { ServicesConstants } from 'services/services.constants';
+import { Stopwatch } from 'domain/utilities/stopwatch.model';
+
+class CyclicStateTransitionsTracker {
+  /** A path of visited states without any repeats. */
+  private pathOfVisitedStates: string[];
+  /** The most recently discovered cycle of visited states. */
+  private cycleOfVisitedStates: string[];
+  private numLoops: number;
+
+  constructor(initStateName: string) {
+    this.pathOfVisitedStates = [initStateName];
+    this.cycleOfVisitedStates = [] ;

```

Comment:
Why is this changed?

Reply:
Its type was string array but here it was assigned null .I

Reply:
```suggestion
    this.cycleOfVisitedStates = [];

```